narrative: how to build variant wheels, build backend implementation (#9)
wheel-next/metadata_pep_devenv: docs/narratives/build_backends.md

   retrieve what that means in terms of `CFLAGS`/`CXXFLAGS`/`LDFLAGS` (that's
   all one can probably support inside the plugin).
2. Pass on those flags, either by amending those environment variables *or* by
   translating them to CLI flags (e.g., C compiler flags can be passed like
   `-Dc_args=xxx`, which is nicer for logging than amending environment
   variables).
3. Once the build is complete and the wheel is being assembled, modify the
   wheel filename and modify the `METADATA` file to insert the variant content

Now let's look at a more complicated example: building a NumPy wheel for
`psabi=x86-64-v3`. The difference is that NumPy contains SIMD code and custom
build flags to control whether and how that code gets included and used.
The build backend cannot know this, hence the user invoking the build must
instruct `meson-python` to only handle step (3) above, and not attempt to do
(1) and (2). This is done by using the `-Cvariant-name` config setting, rather
than `-Cvariant`:

```bash
# Build a NumPy psabi=x86-64-v3 wheel; this requires nonstandard flags (we want
# to keep >v3 runtime-dispatched features)

> $ python -m build --wheel -Cvariant-name="psabi=x86-64-v3" -Csetup-args=-Dcpu-baseline=AVX2,FMA3 -Csetup-args=-Dcpu-dispatch=AVX512F,AVX512_SKX

Reviewer: Do I understand correctly that:

- `-Dvariant=` means "perform full setup to obtain this variant"
- and `-Dvariant-name=` means "only mark the variant, don't do any setup"?

If yes, does this imply that you're also setting, say, `-march=x86-64-v3` manually, or forfeiting it here?

Author: Yes, that is correct indeed. Let me make that more explicit in the text.

> If yes, does this imply that you're also setting, say, `-march=x86-64-v3` manually, or forfeiting it here?

The numpy way of spelling that is `baseline=AVX2,FMA3` (those are the highest instructions that are supported/used in the numpy build, and part of x64-64-v3), so yes that is done manually here.

Is `-march=x86-64-v3` an actual compiler flag, and if so is it portable?

Reviewer: If I'm reading https://gcc.gnu.org/bugzilla/show_bug.cgi?id=97250 right, they're available since GCC 11, and by `git describe`, Clang 12 (Apr 2021).

But to honest, I feel uneasy about letting plugins mangle CFLAGS.

Author: I explicitly called out now that we have `-Cvariant` and `-Cvariant-name`.

            python -m pip install mkl
            export VARIANT_CONFIG=\"variant-name=mkl setup-args=-Dblas=mkl\ $VARIANT_CONFIG"
          fi
          echo CIBW_CONFIG_SETTINGS=$VARIANT_CONFIG >> "$GITHUB_ENV"

      - name: Build wheels
        uses: pypa/cibuildwheel@7940a4c0e76eb2030e473a5f864f291f63ee879b # v2.21.3
        env:
          CIBW_BUILD: ${{ matrix.python[0] }}-${{ matrix.buildplat[1] }}*
          CIBW_ARCHS: ${{ matrix.buildplat[2] }}
          TODO: pass on build variants!

      - uses: actions/upload-artifact@b4b15b8c7c6ac21ea08fcf65892d2ee8f75cf882 # v4.4.3
        with:
          path: ./wheelhouse/*.whl
          name: ${{ matrix.python[0] }}-${{ matrix.buildplat[1] }}
            ${{ matrix.buildplat[2] }} ${{ matrix.buildplat[3] }}
            ${{ matrix.buildplat[4] }}
```

This should yield four wheels named like (plus four more with `cp312` instead of `cp313`):
```
numpy-2.2.4-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
numpy-2.2.4-cp313-cp313-manylinux_2_17_x86_64.musllinux2014_x86_64.whl
numpy-2.2.4-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64_h03981ea44_psabi3.whl

> numpy-2.2.4-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64_h0f4317532_psabi3_mkl.whl

Reviewer 2: You think it's better for the `variant_hash+name` to be at the end ?

The current design has it slightly different: https://mockhouse.wheelnext.dev/pep-xxx-variants/dummy-project/

I like your idea of `+alias` I would use the `+` character :)

Author: Let's defer that to gh-5? @mgorny has already thought about it harder than I have. Here the only key bit is we need the strings in addition to the hash. I don't mind too much where it goes, at least for the purposes of this build backend focused exercise.

Reviewer 2: Agreed - it fundamentally doesn't change anything

Reviewer: Yeah, let's discuss it there. However, note that as it is stated now, it will be interpreted as "platform tag" of (`manylinux_2_17_x86_64` OR `manylinux2014_x86_64_h03981ea44_psabi3`), so a non-variant-enabled PM will consider all versions suitable for the former platform.

Author: Good catch. I changed this to match https://github.com/wheelnext/mockhouse/pull/4 now, so I'll resolve this comment.

# Building wheel variants - UX design

*Primary target audience: developers who want to build wheels for distribution
on PyPI or another index server.*

## Querying plugin variables and values

Before we start building, we need to know what variants can potentially be built.
For now we assume that a plugin provides:

> 1. Variables (e.g., `blas`, `mpi`, `threading`, `cuda`, `rocm`, `psabi`, `simd`,)
> 2. Values (the known set of values for each variable)

Reviewer: ```suggestion
1. Features (e.g., `blas`, `mpi`, `threading`, `cuda`, `rocm`, `psabi`, `simd`,)
2. Values (the known set of values for each feature)
```

to sync the nomenclature.

This may look something like this:

```bash
$ cuda-plugin --list-variables
cuda
...
$ cuda-plugin --list-values cuda
11.6
12.0
12.4
12.6

$ x86-plugin --list-variables
psabi

$ x86-plugin --list-values psabi
x86-64-v1
x86-64-v2
x86-64-v3
x86-64-v4
```

However the details of the UX for the plugin itself don't matter too much for
our purposes in this document - let's assume that other tools have access to
these variables and values, which are determined with some combination of the
plugin's built-in functionality and user-provided configuration.

## Telling build backend to build a particular variant

A wheel variant build is invoked through a build frontend (e.g., `pip`, `build`,
`uv build`). There's multiple possible ways for the actual build (dependencies
linked against, compiler flags, etc.) to be influenced:

> 1. The plugin may set some standard environment variables (e.g., `CFLAGS`,

Reviewer 3: I absolutely love the idea of standardizing the environment variables / build flags that are passed through this variant system.

It should make it a lot easier for people to build things.

Reviewer 3: One other question I have is how do we know which plugins should be used here?

Is this something that would need to be defined in the `pyproject.toml`'s `build-backend section like so?

```toml
[build-system]
requires = ["hatchling", "x86-plugin"]
build-backend = "hatchling.build"
```

Or would we add onto this system with an extra variant influenced section?

```toml
[build-system]
requires = ["hatchling", "x86-plugin"]

[build-system.variant.psabi]
requires = ["x86-plugin"]
```

Author: That's a good question - and I think an important open issue for the whole design. It matters less for the build backend story than for the package install story. I think `[build-system]` is not the right place, since the runtime and build-time plugins are the same. They probably should be defined in metadata separately from the existing build and runtime dependencies.

The bigger question is still opt-in vs. default; I think we're leaning to opt-in (or maybe configurable?). So perhaps something like:
```ini
[variant-wheels]
plugins = [
    "x86-plugin"; platform_machine=='x86_64'",
    "blas-plugin>0.3.0",
]
```

That gives a few opportunities like:
- For setting up an isolated build env, the frontend only has to include plugins if `config_settings` is not empty (or even only if it contains the string "variant")
- For runtime dependencies, the frontend can decide to offer a UX to the user to include or exclude plugins by default or not, rather than make it an always/never choice based on whether we put them in `project.dependencies` yes or no.

Reviewer: Another option is something like the following (which is what I'm experimenting with):

```toml
[build-system]
build-backend = "mesonpy"
requires = [
    "meson-python",
    "variant_x86_64 ; 'x86_64' in variants"
]
```

i.e. using environment specifiers to determine whether we need a particular plugin for a given build.

For installing wheels, the relevant package specifications are included in the metadata (as `Variant-Provider` fields).

Author: Ah I missed replying to this. I see that you implemented support this last option in https://github.com/wheelnext/variantlib/pull/32. My sense is that it's going to be duplicate if it is added to `build-system.requires`.

> For installing wheels, the relevant package specifications are included in the metadata (as Variant-Provider fields).

This may work, or may give extra issues because it's dynamic. Unless you have _all_ `Variant-Provider` entries for all possible wheels listed - but where does the build backend get them from then? I think it works better if they're all declared in `pyproject.toml`; if they're mixed with regular dependencies, I don't see how the build system can know what the set of variant plugins is.

Author: Also a rollout consideration: a top-level `[variant-wheels]` can be used on day 1, while a new environment marker takes at least a year to roll out (all packaging tools must implement and release support for it first, then users must update to the latest version of those tools).

Reviewer: Rollout is a good point. However, to answer your earlier question:

> Unless you have all Variant-Provider entries for all possible wheels listed - but where does the build backend get them from then?

If you're building a specific wheel variant, the plugin needs to be installed to verify that the variant description is correct. In that case, `variantlib` is able to map the loaded plugin into the package that provided it, and therefore include all `Variant-provider` entries relevant to the specific variant. We then combine the entries for all variants when constructing `variants.json`.

Author: The syntax question about how to specify dependencies is solved, the current state of the design looks good.

> standardizing the environment variables / build flags

That's also covered by gh-23 (which has questions, not answers). Let me try and answer there.

> If you're building a specific wheel variant, the plugin needs to be installed to verify that the variant description is correct. In that case, `variantlib` is able to map the loaded plugin into the package that provided it, and therefore include all `Variant-provider` entries relevant to the specific variant. We then combine the entries for all variants when constructing `variants.json`.

This is the problem I was asking about I think, not the solution. You're also saying "it's dynamic" here. Given the current design, they are all listed in `pyproject.toml` though, so that metadata can all be transferred to `METADATA` in wheels.

   `CXXFLAGS`, `LDFLAGS`).
2. The build backend may recognize a variant variable plus value and do
   something specific to that (e.g., try detecting and linking a dependency).
3. The user building the wheel may take actions to ensure the build does the
   desired things to produce the wheel variant, e.g.:

   - Provide the variant name and value
   - Install required build-time dependencies before starting the build
   - Apply patches to the package's source code before starting the build
   - Modify environment variables
   - Pass custom build flags to the backend via `--config-settings`
   - Post-process the wheel, either with a tool like `auditwheel` or manually

The simplest case is only (1). Invoking such a build for a package that respects
`CFLAGS`/`LDFLAGS` and needs nothing else could look like:

```bash
# E.g., a default psabi-v3 usage with meson-python

> $ python -m build -Cvariant="psabi=x86-64-v3"

Reviewer: ```suggestion
$ python -m build -Cvariant="x86_64::level::v3"
```

```

What actually happens under the hood for that build command? The build frontend,
`build` in this case, only passes on the `-C` (shorthand for `--config-settings`)
to the build backend. The build backend, `meson-python` in this case, may do the
following:

> 1. Query the plugins, find that `x86-plugin` provides the `psabi` variable, then

Reviewer: ```suggestion
1. Query the plugins, find that `variant_x86_64` provides the `x86_64 :: psabi` feature, then
```

   retrieve what that means in terms of `CFLAGS`/`CXXFLAGS`/`LDFLAGS` (that's
   all one can probably support inside the plugin).
2. Pass on those flags, either by amending those environment variables *or* by
   translating them to CLI flags (e.g., C compiler flags can be passed like
   `-Dc_args=xxx`, which is nicer for logging than amending environment
   variables).
3. Once the build is complete and the wheel is being assembled, modify the
   wheel filename and modify the `METADATA` file to insert the variant content

Now let's look at a more complicated example: building a NumPy wheel for

> `psabi=x86-64-v3`. The difference is that NumPy contains SIMD code and custom

Reviewer: ```suggestion
`x86_64 :: level :: v3`. The difference is that NumPy contains SIMD code and custom
```

build flags to control whether and how that code gets included and used.
The build backend cannot know this, hence the user invoking the build must
instruct `meson-python` to only handle step (3) above, and not attempt to do
(1) and (2). This is done by using the `-Cvariant-name` config setting, rather
than `-Cvariant`:

```bash

> # Build a NumPy psabi=x86-64-v3 wheel; this requires nonstandard flags (we want

Reviewer: ```suggestion
# Build a NumPy x86_64::level::v3 wheel; this requires nonstandard flags (we want
```

# to keep >v3 runtime-dispatched features)

> $ python -m build --wheel -Cvariant-name="psabi=x86-64-v3" -Csetup-args=-Dcpu-baseline=AVX2,FMA3 -Csetup-args=-Dcpu-dispatch=AVX512F,AVX512_SKX

Reviewer: ```suggestion
$ python -m build --wheel -Cvariant-name="x86_64::level::v3" -Csetup-args=-Dcpu-baseline=AVX2,FMA3 -Csetup-args=-Dcpu-dispatch=AVX512F,AVX512_SKX
```

Also, technically we've actually solved this via `-Cvariant`, with the variant being passed to setup args, so perhaps we ought to use a different example here.

```

Note that we've introduced two ways of specifying a variant build so far (which
is probably all we'll need):

> - `-Cvariant=variable=value`: specify the wheel variant to build *and* let

Reviewer: ```suggestion
- `-Cvariant=namespace::feature::value`: specify the wheel variant to build *and* let
```

  the plugin add defaults (compile and link flags)

> - `-Cvariant-name=variable=value`: only specify the wheel variant to build

Reviewer: ```suggestion
- `-Cvariant-name=namespace::feature::value`: only specify the wheel variant to build
```

So far so good for how to build x86-64 SIMD variants. Let's look at a second
realistic case, namely BLAS variants. The difference is that we need not only
compile and link flags, but also a different dependency - during the build,
the correct shared library must be found *and* before the build, the user
must have prepared the build environment correctly to ensure the shared library
is actually present.

Given the different dependencies here, and the relevant linker flags being
different per build system, the plugin has no way to provide any compile/link
flags here or do anything else that's helpful (it doesn't know after all which
build backend we'll be using).

Here is what the plugin's variables and values may be:

```bash
$ blas-plugin --list-variables
blas

> lapack

Reviewer: Do we ever want to use a different lapack provider than blas?

Author: Yes, for example BLIS is a BLAS-only library, so you'd use it with Netlib LAPACK or with libFlame. OpenBLAS can also be built as BLAS-only.
